### ISSUE TITLE
Any way to do multisig with Ledger via HWI?

### ISSUE THREAD

**Issue by mflaxman** [bitcoin-core/HWI#596]
It seems that there is code that supports it here:
https://github.com/bitcoin-core/HWI/tree/master/hwilib/devices/ledger_bitcoin

I'm looking at [the related PR](https://github.com/bitcoin-core/HWI/pull/550) and @achow101 says "I've implemented support for multisigs." Is there somewhere that shows this working / how this works? I can't find it.

I got a full PoC working of exporting an xpub, registering a wallet (collection of xpubs, quorum, script type, etc), showing an address on device, and then (after depositing funds) signing a transaction using the Ledger code that this appears based off of.
https://gist.github.com/mflaxman/386990be297e9a595936e8142e815e52

Basically, I'm trying to decide if I want to package/publish what I've done into some open source code that allows end-users to user their ledger for p2wsh transactions. I'm wondering:
1. Should I use hwilib for the dependency or `from hwilib.devices.ledger_bitcoin import MultisigWallet` vs `from ledger_bitcoin import MultisigWallet`. My default assumption is that `hwilib` would be a better dependency, but if it's an unsupported feature perhaps I should pull from upstream.
1. Does that really not exist? [My gist above](https://gist.github.com/mflaxman/386990be297e9a595936e8142e815e52) may be the only ledger p2wsh transactions to ever hit the bitcoin blockchain, but considering [this feature was announced by Ledger ~5 months ago](https://blog.ledger.com/bitcoin-2/) that seems hard to believe.

**Comment by achow101** [comment]
Yes, it is. What have you tried to do that does not work?

With the Ledger 2.0 app, in order to sign a multisig the policy needs to be registered on the device first, and then it will send back an HMAC that the client needs to present each time it wants to use that policy. However HWI does not have a way to store this HMAC nor does it have a way to register a wallet with the common interface. What HWI will do is register the policy each time observes a multisig policy. This means that from a user prospective, they will have to confirm the policy each time an action such as signing a transaction is done. This means that you can do multisigs with just descriptors and PSBTs without needing to explicitly register a policy or go into the device specific details.

With the Ledger 1.0 (legacy) app, all of this is possible, just without the need to register a wallet policy.

It should be noted that by using `hwilib.devices.ledger_bitcoin` that you are using an internal API that is not meant for external use. I suppose it should be prefixed with an underscore.

**Comment by mflaxman** [comment]
> With the Ledger 1.0 (legacy) app, all of this is possible, just without the need to register a wallet policy.

Wait, ledger 1.0 supports multisig!? My understanding was that there was no way to display a multisig address on device, and signing couldn't detect change (effectively making it blind-signing for all but the most expert of users). That's how Electrum's implementation worked, and I had heard that it was a technical limitation that only ledger 2.0 finally enabled so I didn't dig into other implementations too closely.

I guess I'm still stuck on the question: is there a way a non-expert could send/receive using a ledger as part of a multisig transaction today (using HWI or something else) in a way that wasn't terribly insecure? I'm baffled that it doesn't appear possible outside of the gist I threw together. If you could provide a link to something showing otherwise it'd be greatly appreciated.

**Comment by achow101** [comment]
I suppose our definitions of "do multisig" slightly differ. Both versions of the Ledger Bitcoin app can sign multisigs. Ledger 2.0 will additionally display information about the multisig it is signing.

In both cases, you can use basically the same workflow with HWI and a you will get a signed multisig transaction. With both versions, you can use HWI's `getxpub` to get a xpub that you can use in your multisig. With both versions, you can provide a PSBT to `signtx` and it will return a PSBT with signatures for the multisig. With version 2, when you do `signtx`, HWI will figure out the multisig policies in the PSBT, register them with the device (which will require user interaction to confirm the registrattion), and then the device should be able to detect change for that policy.

Displaying a multisig address on Ledgers is not yet implemented.

**Comment by mflaxman** [comment]
> I suppose our definitions of "do multisig" slightly differ.

Ya, I'm talking about:
1. Validating a receive address on the screen of your trusted display before giving it out to anyone for payment
2. Being able to tell what you're signing without blind signing (anything beyond "send X bitcoin to Y address with fee of Z" is too complicated for non-experts)

I feel strongly that it's a terrible idea to ever use a hardware that doesn't natively offer these features.

With that gist I threw together it seems ledger now supports multisig in a way that isn't dangerously insecure, there just isn't a UI yet to make it available to non-experts. Perhaps I'll work on that.

Closing this issue, I'll use other software to accomplish the goal. FWIW, I do think HWI adding support for secure multisig with ledger would be a great feature addition. In the meantime, I'd consider adding a lot of big scary warnings.

FYI: I didn't audit your signing flow (I couldn't validate a receive address first), but registering a new quorum on each sign has obvious dangers from a compromised host machine. Stateless Signer trustlessly validate change UTXOs already (Trezor first did this 6+ years ago), quorum registration is about receive address validation ONLY, there's no reason to do be doing that on signing (and it introduces big problems).

**Comment by achow101** [comment]
> but registering a new quorum on each sign has obvious dangers from a compromised host machine. Stateless Signer trustlessly validate change UTXOs already (Trezor first did this 6+ years ago), quorum registration is about receive address validation ONLY, there's no reason to do be doing that on signing (and it introduces big problems).

HWI does it this way because it is stateless. It cannot remember the HMAC that the ledger returns for use in other commands, so it must do the policy registration each time it does something with the policy. Trezor works because it is stateful and it has the storage to persist the policy, while Ledgers do not have the storage and are stateless and so require the state to be client side. Unfortunately that does not work when the client side is also stateless as HWI is.

**Comment by mflaxman** [comment]
> Trezor works because it is stateful and it has the storage to persist the policy

I think maybe you're thinking of single sig? Trezor has been detecting change outputs [statelessly from the transaction since 2015](https://twitter.com/pavolrusnak/status/1166375789534863362). Now with PSBT (HT to you!) this is even easier, [my stateless CLI multisig wallet](https://twitter.com/mflaxman/status/1321503036724989952) validates change by calculating change outputs use the same xpubs/quorum as the inputs. You don't need to prompt the user to confirm something confusing; the device can cryptographically validate.

To be clear: I'm not saying that HWI should be stateful.

**Comment by craigraw** [comment]
> It cannot remember the HMAC that the ledger returns for use in other commands, so it must do the policy registration each time it does something with the policy.

Does it make sense for HWI to return the HMAC as an additional JSON field in those calls where the wallet policy is registered, and to accept it as an optional parameter for future calls that can use it to avoid repeat registration? Wallet software can then offer the stateful backing HWI lacks.

**Comment by hodlwave** [comment]
@achow101 would you mind responding to the [last comment](https://github.com/bitcoin-core/HWI/issues/596#issuecomment-1124637176) from @craigraw?

**Comment by achow101** [comment]
> Does it make sense for HWI to return the HMAC as an additional JSON field in those calls where the wallet policy is registered, and to accept it as an optional parameter for future calls that can use it to avoid repeat registration? Wallet software can then offer the stateful backing HWI lacks.

I don't think it makes sense to do that as part of `signtx`, however it could make sense as a separate policy registration command if other wallets also support a similar workflow. I don't think it's useful to implement a feature that only one kind of device supports.

**Comment by bitcoin-eagle** [comment]
> policy registration command

the other devices do implicit policy registration at any point (signing or checking address) and it works because they have a memory. whereas devices without memory need to store it externally in the memory of a computer. Devices like jade or coldcard already have policy registration workflow triggered implicitly. So IMHO we should support explicit policy registration to cover all the cases even for devices without memory; otherwise devices without memory will always be second class citizens
